docs/architecture/adr-038-add-support-for-wasm-light-clients.md

```diff
@@ -0,0 +1,86 @@
+# ADR 38: Add support for Wasm based light client
+
+## Changelog
+
+- 26/11/2020: Initial Draft
+
+## Status
+
+Draft
+
+## Abstract
+
+Currently in the SDK, light clients are part of Cosmos SDK codebase. What this means that, anytime a blockchain built 
+upon Cosmos SDK need to add support for new light client, Cosmos SDK need to be modified and all validator nodes of that 
+blockchain need to be updated with custom version of Cosmos SDK. 
+
+To remedy these shortcomings, we are proposing a WASM VM to host light client bytecode, which allows easier upgrading of 
+existing light clients as well as adding support for new light clients without requiring a code release and corresponding 
+hard-fork event.
+
+## Context
+Currently in the SDK, light clients are defined as part of the codebase and are implemented as submodules under 
+`/x/ibc/light-clients/`. To add support for new light client or update an existing light client in the event of security 
+issue or consensus update, we need to modify the codebase and integrate it in *numerous* places. In addition, individual 
+blockchains built upon Cosmos SDK need to update all their validator nodes to latest version to add support for this 
+light client. This entire process is tedious and time consuming. In the event a blockchain wants to add support for niche 
+light client, it need to fork Cosmos SDK and modify codebase to add support for it. This creates overhead of maintaining 
+that fork against mainstream Cosmos SDK release.
```

Comment:
I think there are two different issues:
1. updating light client in when doing a chain breaking change (eg consensus) - this one will need a hard fork and a network update  anyway.
2. doing iterative light client update or adding a new light client

In this ADR we are targeting the second use-case. Right? If so, let's make it clear.

Reply:
I don't believe this is quite the case:

In the context of this ADR, (1) only requires a hard fork of the local chain in the instance that it's own consensus changes. 

As it stands currently, with a Tendermint light client that is tightly coupled to the local chain's own version of Tendermint (ergo, ICS-007), any Tendermint chain that wants to connect to the Cosmos HUB, for example, has to maintain parity with, or at least run a compatible version of the light client, and vice versa. This means that should Tendermint change materially (I'm not sure whether the changes between TM 0.33 and 0.34 are thusly affected, but for this purpose lets assume that they are), then you could not connect Chain A running 0.33 and Chain B running 0.34 - in either direction; all interconnected chains within the ecosystem would need to run the same version, and organising multiple concurrent (or near-concurrent) upgrades isn't feasible.

Reply:
Correct. This makes sense. Let's include @joe-bowman context in the ADR.